core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/Utils.scala

```diff
@@ -45,8 +46,30 @@ object Utils {
   }
   val AMBER_HOME_FOLDER_NAME = "amber";
 
+  /**
+    * Retry the given logic with a backoff time interval. The attempts are executed sequentially, thus blocking the thread.
+    * Backoff time is doubled after each attempt.
+    * @param attempts total number of attempts. if n <= 1 then it will not retry at all, decreased by 1 for each recursion.
+    * @param baseBackoffTimeInMS time to wait before next attempt, started with the base time, and doubled after each attempt.
+    * @param fn the target function to execute.
+    * @tparam T any return type from the provided function fn.
+    * @return the provided function fn's return, or any exception that still being raised after n attempts.
+    */
+  @tailrec
+  def retry[T](attempts: Int, baseBackoffTimeInMS: Long)(fn: => T): T = {
+    try {
+      fn
+    } catch {
+      case e: Throwable =>
+        if (attempts > 1) {
+          println("retrying after " + baseBackoffTimeInMS + "ms")
```

Comment:
should we also print the error message / stack trace here? it's good to at least see it in the console output on why it is retrying.
Also maybe add "attempt n" in print message

Reply:
sure